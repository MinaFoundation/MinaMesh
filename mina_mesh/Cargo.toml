[package]
name = "mina_mesh"
version.workspace = true
edition = "2021"

[lib]
path = "mod.rs"

[dependencies]
anyhow = "1.0.86"
cynic = { version = "3.7.3", features = ["http-reqwest-blocking"] }
envy = "0.4.2"
reqwest = { version = "0.12.5", features = ["json", "blocking"] }
serde = { version = "1.0.204", features = ["derive"] }
serde_json = { version = "1.0.121" }
sqlx = { version = "0.8.0", features = ["runtime-tokio", "postgres", "json"] }
thiserror = "1.0.63"
tokio = { version = "1.39.2", features = ["full", "test-util"] }

mesh = { path = "../mesh_generated" }
mina_mesh_graphql = { path = "../mina_mesh_graphql" }
