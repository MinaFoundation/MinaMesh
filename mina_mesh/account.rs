use crate::graphql_generated::mina::{
  Account, AnnotatedBalance, Balance, Length, PublicKey, QueryBalance, QueryBalanceVariables, StateHash,
};
use crate::MinaMeshContext;
use anyhow::Result;
use cynic::{http::ReqwestExt, QueryBuilder};
use mesh::models::{
  AccountBalanceRequest, AccountBalanceResponse, AccountIdentifier, Amount, BlockIdentifier, Currency,
  PartialBlockIdentifier,
};

/// https://github.com/MinaProtocol/mina/blob/985eda49bdfabc046ef9001d3c406e688bc7ec45/src/app/rosetta/lib/account.ml#L11
pub async fn balance(request: AccountBalanceRequest) -> Result<AccountBalanceResponse> {
  let context = MinaMeshContext::from_env().await?;
  context.network_health_check(*request.network_identifier).await?;
  let account: MinaAccountIdentifier = (*request.account_identifier).into();
  match request.block_identifier {
    Some(block_identifier) => block_balance(&context, &account, *block_identifier).await,
    None => frontier_balance(&context, &account).await,
  }
}

const DEFAULT_TOKEN_ID: &str = "wSHV2S4qX9jFsLjQo8r1BsMLH2ZRKsZx6EJd1sbozGPieEC4Jf";

async fn block_balance(
  context: &MinaMeshContext,
  account_identifier: &MinaAccountIdentifier,
  block_identifier: PartialBlockIdentifier,
) -> Result<AccountBalanceResponse> {
  // Get block data from the database
  let maybe_block = sqlx::query_file!("sql/maybe_block.sql", block_identifier.index)
    .fetch_optional(&context.pool)
    .await?;
  match maybe_block {
    Some(block) => {
      // has canonical height
      // do we really need to do a different query?
      let maybe_account_balance_info = sqlx::query_file!(
        "sql/maybe_account_balance_info.sql",
        account_identifier.public_key,
        block_identifier.index.unwrap(),
        account_identifier.token_id,
      )
      .fetch_optional(&context.pool)
      .await?;
      match maybe_account_balance_info {
        None => {
          return Ok(AccountBalanceResponse::new(
            BlockIdentifier {
              hash: block.state_hash,
              index: block.height,
            },
            vec![Amount {
              currency: Box::new(Currency {
                symbol: "MINA".into(), // TODO: Use actual currency symbol / custom tokens
                decimals: 9,
                metadata: None,
              }),
              value: "0".to_string(),
              metadata: Some(serde_json::json!({
                "locked_balance": "0".to_string(),
                "liquid_balance": "0".to_string(),
                "total_balance": "0".to_string()
              })),
            }],
          ));
        }
        Some(account_balance_info) => {
          println!("B");
          let last_relevant_command_balance = account_balance_info.balance.parse::<u64>()?;
          let timing_info = sqlx::query_file!("sql/timing_info.sql", account_balance_info.timing_id)
            .fetch_optional(&context.pool)
            .await?;
          let liquid_balance = match timing_info {
            Some(timing_info) => {
              let incremental_balance = incremental_balance_between_slots(
                account_balance_info.block_global_slot_since_genesis as u32,
                block.global_slot_since_genesis as u32,
                timing_info.cliff_time as u32,
                timing_info.cliff_amount.parse::<u64>().unwrap(),
                timing_info.vesting_period as u32,
                timing_info.vesting_increment.parse::<u64>().unwrap(),
                timing_info.initial_minimum_balance.parse::<u64>().unwrap(),
              );
              last_relevant_command_balance + incremental_balance
            }
            None => last_relevant_command_balance,
          };
          let total_balance = last_relevant_command_balance;
          let locked_balance = total_balance - liquid_balance;
          Ok(AccountBalanceResponse::new(
            BlockIdentifier {
              hash: block.state_hash,
              index: block.height,
            },
            vec![Amount {
              currency: Box::new(Currency {
                symbol: "MINA".into(), // TODO: Use actual currency symbol / custom tokens
                decimals: 9,
                metadata: None,
              }),
              value: liquid_balance.to_string(),
              metadata: Some(serde_json::json!({
                "locked_balance": locked_balance.to_string(),
                "liquid_balance": liquid_balance.to_string(),
                "total_balance": total_balance.to_string()
              })),
            }],
          ))
        }
      }
    }
    None => {
      // TODO: return proper error message
      anyhow::bail!("Block not found")
    }
  }
}

fn min_balance_at_slot(
  global_slot: u32,
  cliff_time: u32,
  cliff_amount: u64,
  vesting_period: u32,
  vesting_increment: u64,
  initial_minimum_balance: u64,
) -> u64 {
  if global_slot < cliff_time {
    initial_minimum_balance
  } else if vesting_period == 0 {
    0
  } else {
    let min_balance_past_cliff = initial_minimum_balance.saturating_sub(cliff_amount);
    if min_balance_past_cliff == 0 {
      0
    } else {
      let num_periods = (global_slot - cliff_time) / vesting_period;
      let vesting_decrement = if (u64::MAX / num_periods as u64) < vesting_increment {
        u64::MAX
      } else {
        num_periods as u64 * vesting_increment
      };
      min_balance_past_cliff.saturating_sub(vesting_decrement)
    }
  }
}

fn incremental_balance_between_slots(
  start_slot: u32,
  end_slot: u32,
  cliff_time: u32,
  cliff_amount: u64,
  vesting_period: u32,
  vesting_increment: u64,
  initial_minimum_balance: u64,
) -> u64 {
  if end_slot <= start_slot {
    return 0;
  }
  let min_balance_at_start_slot = min_balance_at_slot(
    start_slot,
    cliff_time,
    cliff_amount,
    vesting_period,
    vesting_increment,
    initial_minimum_balance,
  );
  let min_balance_at_end_slot = min_balance_at_slot(
    end_slot,
    cliff_time,
    cliff_amount,
    vesting_period,
    vesting_increment,
    initial_minimum_balance,
  );
  min_balance_at_start_slot.saturating_sub(min_balance_at_end_slot)
}

// Note: The `min_balance_at_slot` function is not provided in the original OCaml code,
// so we'll declare it here as a separate function that needs to be implemented.

async fn frontier_balance(
  context: &MinaMeshContext,
  address: &MinaAccountIdentifier,
) -> Result<AccountBalanceResponse> {
  let operation = QueryBalance::build(QueryBalanceVariables {
    public_key: PublicKey(address.public_key.clone()),
  });
  let result = context
    .client
    .post(&context.config.mina_proxy_url)
    .run_graphql(operation)
    .await?;
  if let Some(QueryBalance {
    account:
      Some(Account {
        balance:
          AnnotatedBalance {
            block_height: Length(index_raw),
            state_hash: Some(StateHash(hash)),
            liquid: Some(Balance(liquid_raw)),
            total: Balance(total_raw),
          },
        ..
      }),
  }) = result.data
  {
    let total = total_raw.parse::<u64>()?;
    let liquid = liquid_raw.parse::<u64>()?;
    let index = index_raw.parse::<i64>()?;
    return Ok(AccountBalanceResponse::new(
      BlockIdentifier { hash, index },
      vec![Amount {
        currency: Box::new(Currency {
          symbol: "MINA".into(), // TODO: Use actual currency symbol / custom tokens
          decimals: 9,
          metadata: None,
        }),
        value: total_raw,
        metadata: Some(serde_json::json!({
          "locked_balance": (total - liquid).to_string(),
          "liquid_balance": liquid.to_string(),
          "total_balance": total.to_string()
        })),
      }],
    ));
  } else {
    anyhow::bail!("Failed to get balance: {:?}", result.errors.or(Some(vec![])))
  }
}

#[derive(Debug)]
pub struct MinaAccountIdentifier {
  pub public_key: String,
  pub token_id: String,
}

impl Into<MinaAccountIdentifier> for AccountIdentifier {
  fn into(self) -> MinaAccountIdentifier {
    let token_id = match self.metadata {
      Some(serde_json::Value::Object(map)) => map.get("token_id").map(|v| v.as_str().unwrap().to_string()),
      None => Some(DEFAULT_TOKEN_ID.to_string()),
      _ => unimplemented!(),
    }
    .unwrap(); // TODO: handle unwrap
    MinaAccountIdentifier {
      public_key: self.address,
      token_id,
    }
  }
}

#[tokio::test]
async fn first() {
  use mesh::models::NetworkIdentifier;
  let x = balance(AccountBalanceRequest {
    account_identifier: Box::new(AccountIdentifier {
      // address: "B62qrQKS9ghd91shs73TCmBJRW9GzvTJK443DPx2YbqcyoLc56g1ny9".into(),
      // address: "B62qooos8xGyqtJGpT7eaoyGrABCf4vcAnzCtxPLNrf26M7FwAxHg1i".into(),
      address: "B62qmjJeM4Fd4FVghfhgwoE1fkEexK2Rre8WYKMnbxVwB5vtKUwvgMv".into(),
      sub_account: None,
      metadata: None,
    }),
    block_identifier: Some(Box::new(PartialBlockIdentifier {
      index: Some(371513),
      hash: None,
    })),
    currencies: None,
    network_identifier: Box::new(NetworkIdentifier {
      blockchain: "mina".into(),
      network: "mainnet".into(),
      sub_network_identifier: None,
    }),
  })
  .await;
  println!("{:?}", x);
}

pub fn coins() {}
