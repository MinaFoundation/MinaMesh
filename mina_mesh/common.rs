use anyhow::{Context, Result};
pub use mesh::models::{AccountIdentifier, NetworkIdentifier};
use mina_mesh_graphql::GraphQLClient;
use serde::Deserialize;
use sqlx::PgPool;

#[derive(Debug)]
pub struct MinaMesh {
  pub env: MinaMeshEnv,
  pub graphql_client: GraphQLClient,
  pub pool: PgPool,
}

impl MinaMesh {
  pub async fn from_env() -> Result<Self> {
    let config = envy::from_env::<MinaMeshEnv>().context("Failed to parse config from env")?;
    let database_url = config.mina_archive_database_url.to_owned();
    let mina_proxy_url = config.mina_graphql_url.to_owned();
    Ok(Self {
      env: config,
      graphql_client: GraphQLClient::new(mina_proxy_url),
      pool: PgPool::connect(database_url.as_str()).await?,
    })
  }
}

#[derive(Deserialize, Debug, Default)]
pub struct MinaMeshEnv {
  #[serde(default = "default_mina_graphql_url", rename(deserialize = "MINA_GRAPHQL_URL"))]
  mina_graphql_url: String,
  #[serde(
    default = "default_mina_archive_database_url",
    rename(deserialize = "MINA_ARCHIVE_DATABASE_URL")
  )]
  mina_archive_database_url: String,
  #[serde(default = "default_genesis_block_identifier_height")]
  pub genesis_block_identifier_height: i64,
  #[serde(default = "default_genesis_block_identifier_state_hash")]
  pub genesis_block_identifier_state_hash: String,
}

fn default_mina_graphql_url() -> String {
  "https://mainnet.minaprotocol.network/graphql".to_string()
}

fn default_mina_archive_database_url() -> String {
  "postgres://mina:whatever@localhost:5432/archive".to_string()
}

fn default_genesis_block_identifier_height() -> i64 {
  359605
}

fn default_genesis_block_identifier_state_hash() -> String {
  "3NK4BpDSekaqsG6tx8Nse2zJchRft2JpnbvMiog55WCr5xJZaKeP".to_string()
}
