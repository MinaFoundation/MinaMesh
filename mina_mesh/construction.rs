/// https://github.com/MinaProtocol/mina/blob/985eda49bdfabc046ef9001d3c406e688bc7ec45/src/app/rosetta/lib/construction.ml#L162
pub fn derive() {}

/// https://github.com/MinaProtocol/mina/blob/985eda49bdfabc046ef9001d3c406e688bc7ec45/src/app/rosetta/lib/construction.ml#L392
pub fn preprocess() {}

/// https://github.com/MinaProtocol/mina/blob/985eda49bdfabc046ef9001d3c406e688bc7ec45/src/app/rosetta/lib/construction.ml#L133
pub fn metadata() {}

/// https://github.com/MinaProtocol/mina/blob/985eda49bdfabc046ef9001d3c406e688bc7ec45/src/app/rosetta/lib/construction.ml#L473
pub fn payloads() {}

/// https://github.com/MinaProtocol/mina/blob/985eda49bdfabc046ef9001d3c406e688bc7ec45/src/app/rosetta/lib/construction.ml#L615
pub fn parse() {}

/// https://github.com/MinaProtocol/mina/blob/985eda49bdfabc046ef9001d3c406e688bc7ec45/src/app/rosetta/lib/construction.ml#L561
pub fn combine() {}

/// https://github.com/MinaProtocol/mina/blob/985eda49bdfabc046ef9001d3c406e688bc7ec45/src/app/rosetta/lib/construction.ml#L786
pub fn hash() {}

/// https://github.com/MinaProtocol/mina/blob/985eda49bdfabc046ef9001d3c406e688bc7ec45/src/app/rosetta/lib/construction.ml#L849
pub fn submit() {}
