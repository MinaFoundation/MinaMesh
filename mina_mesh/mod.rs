pub mod account;
mod block;
pub mod construction;
mod mempool;
pub mod network;

use crate::graphql_generated::mina::QueryNetworkId;
use anyhow::{bail, Context as AnyhowContext, Result};
use cynic::{http::ReqwestExt, QueryBuilder};
use mesh::models::NetworkIdentifier;
use reqwest::Client;
use serde::Deserialize;
use sqlx::PgPool;
use std::vec::Vec;

#[derive(Deserialize, Debug, Default)]
pub struct Config {
  #[serde(default = "default_mina_proxy_url")]
  mina_proxy_url: String,
  #[serde(default = "default_archive_url")]
  archive_url: String,
  #[serde(default = "default_database_url")]
  database_url: String,
  genesis_block_identifier: String,
}

fn default_mina_proxy_url() -> String {
  "https://mainnet.minaprotocol.network/graphql".to_string()
}

fn default_archive_url() -> String {
  "https://api.minascan.io/archive/devnet/v1/graphql".to_string()
}

fn default_database_url() -> String {
  "postgres://mina:whatever@localhost:5432/archive".to_string()
}

pub struct MinaMeshContext {
  config: Config,
  client: Client,
  pool: PgPool,
}

impl MinaMeshContext {
  async fn from_env() -> Result<Self> {
    let config = envy::from_env::<Config>().with_context(|| "Failed to parse config from env")?;
    let database_url = config.database_url.clone();
    Ok(Self {
      config,
      client: Client::new(),
      pool: PgPool::connect(database_url.as_str()).await?,
    })
  }

  async fn network_health_check(&self, network_identifier: NetworkIdentifier) -> Result<bool> {
    let QueryNetworkId { network_id } = self.graphql(QueryNetworkId::build(())).await?;
    if network_identifier.blockchain == "MINA" {
      unimplemented!();
    }
    if &network_identifier.network == &network_id {
      unimplemented!();
    }
    Ok(true)
  }

  async fn graphql<ResponseData, Vars>(&self, operation: cynic::Operation<ResponseData, Vars>) -> Result<ResponseData>
  where
    Vars: serde::Serialize,
    ResponseData: serde::de::DeserializeOwned + 'static,
  {
    let response = self
      .client
      .post(&self.config.mina_proxy_url)
      .run_graphql(operation)
      .await
      .context("Failed to run GraphQL query")?;
    if let Some(errors) = response.errors {
      bail!(errors
        .iter()
        .map(|err| err.message.clone())
        .collect::<Vec<String>>()
        .join("\n\n"));
    } else if let Some(data) = response.data {
      Ok(data)
    } else {
      bail!("No data contained in GraphQL response");
    }
  }
}

trait ToVecOfString {
  fn to_vec_of_string(self) -> Vec<String>;
}

impl ToVecOfString for Vec<&str> {
  fn to_vec_of_string(self) -> Vec<String> {
    self.into_iter().map(|s| s.to_string()).collect()
  }
}
