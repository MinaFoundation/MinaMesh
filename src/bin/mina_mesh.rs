// TODO: document workflow regarding fetching and using initial genesis ledger
// hash.

use anyhow::Result;
use clap::Parser;
use mina_mesh::{FetchGenesisBlockIdentifierCommand, ServeCommand};

#[derive(Debug, Parser)]
#[command(
  name = "mina-mesh",
  version,
  about = "A Mesh-compliant Server for Mina",
  propagate_version = true,
  author
)]
enum Command {
  #[command(about = "Start the Mina Mesh Server.")]
  Serve(ServeCommand),
  #[command(about = "Retrieve the genesis block identifier via a proxy node GraphQL endpoint.")]
  FetchGenesisBlockIdentifier(FetchGenesisBlockIdentifierCommand),
}

#[tokio::main]
async fn main() -> Result<()> {
  match Command::parse() {
    Command::Serve(cmd) => cmd.execute().await,
    Command::FetchGenesisBlockIdentifier(cmd) => cmd.execute().await,
  }
}
