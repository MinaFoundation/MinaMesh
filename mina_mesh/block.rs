use crate::common::MinaMesh;
use anyhow::Result;
pub use mesh::models::{Block, BlockIdentifier, BlockRequest, BlockResponse};

/// https://github.com/MinaProtocol/mina/blob/985eda49bdfabc046ef9001d3c406e688bc7ec45/src/app/rosetta/lib/block.ml#L7
impl MinaMesh {
  pub async fn block(&self, _request: BlockRequest) -> Result<BlockResponse> {
    Ok(BlockResponse {
      block: Some(Box::new(Block::new(
        BlockIdentifier::new(0, "".to_string()),
        BlockIdentifier::new(0, "".to_string()),
        0,
        vec![],
      ))),
      other_transactions: Some(vec![]),
    })
  }
}
