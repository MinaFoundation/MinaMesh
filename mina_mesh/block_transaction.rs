use crate::common::MinaMeshContext;
use anyhow::Result;
use cynic::QueryBuilder;
use mesh::models::{BlockTransactionRequest, BlockTransactionResponse, Transaction, TransactionIdentifier};
use mina_mesh_graphql::{Block2, QueryBlockTransactions, QueryBlockTransactionsVariables, Transactions};

pub async fn block_transaction(
  context: &MinaMeshContext,
  transaction: &BlockTransactionRequest,
) -> Result<BlockTransactionResponse> {
  let QueryBlockTransactions {
    block: Block2 {
      transactions: Transactions { user_commands },
    },
  } = context
    .graphql(QueryBlockTransactions::build(QueryBlockTransactionsVariables {
      state_hash: Some(transaction.block_identifier.hash.as_str()),
    }))
    .await?;
  Ok(BlockTransactionResponse::new(Transaction {
    operations: user_commands.into_iter().map(|command| command.into()).collect(),
    metadata: None,
    related_transactions: Some(vec![]),
    transaction_identifier: Box::new(TransactionIdentifier::new(
      transaction.transaction_identifier.to_owned().hash,
    )),
  }))
}
