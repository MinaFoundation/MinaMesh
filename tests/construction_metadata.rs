use anyhow::Result;
use insta::assert_debug_snapshot;
use mina_mesh::{models::ConstructionMetadataRequest, test::network_id, MinaMeshConfig};

#[tokio::test]
async fn construction_metadata_ok() -> Result<()> {
  let mina_mesh = MinaMeshConfig::from_env().to_mina_mesh().await?;

  let request = ConstructionMetadataRequest {
    network_identifier: network_id().into(),
    options: Some(serde_json::json!({
      // cspell:disable
      "sender": "B62qkd6yYALkQMq2SFd5B57bJbGBMA2QuGtLPMzRhhnvexRtVRycZWP",
      "receiver": "B62qnXy1f75qq8c6HS2Am88Gk6UyvTHK3iSYh4Hb3nD6DS2eS6wZ4or",
      "token_id": "wSHV2S4qX9jFsLjQo8r1BsMLH2ZRKsZx6EJd1sbozGPieEC4Jf",
      // cspell:enable
      "valid_until": "20000",
      "memo": "test transaction"
    })),
    public_keys: None,
  };

  let response = mina_mesh.construction_metadata(request).await;

  assert!(response.is_ok());
  let response = response.unwrap();
  let metadata_str = serde_json::to_string(&response.metadata)?;
  // cspell:disable
  assert!(metadata_str.contains("B62qkd6yYALkQMq2SFd5B57bJbGBMA2QuGtLPMzRhhnvexRtVRycZWP"));
  assert!(metadata_str.contains("B62qnXy1f75qq8c6HS2Am88Gk6UyvTHK3iSYh4Hb3nD6DS2eS6wZ4or"));
  assert!(metadata_str.contains("wSHV2S4qX9jFsLjQo8r1BsMLH2ZRKsZx6EJd1sbozGPieEC4Jf"));
  // cspell:enable
  assert!(metadata_str.contains("test transaction"));
  assert!(metadata_str.contains("20000"));
  assert!(metadata_str.contains("account_creation_fee"));
  Ok(())
}

#[tokio::test]
async fn construction_metadata_acct_not_found() -> Result<()> {
  let mina_mesh = MinaMeshConfig::from_env().to_mina_mesh().await?;

  let request = ConstructionMetadataRequest {
    network_identifier: network_id().into(),
    options: Some(serde_json::json!({
      // cspell:disable
      "sender": "B62qpLST3UC1rpVT6SHfB7wqW2iQgiopFAGfrcovPgLjgfpDUN2LLeg",
      "receiver": "B62qnXy1f75qq8c6HS2Am88Gk6UyvTHK3iSYh4Hb3nD6DS2eS6wZ4or",
      // cspell:enable
      "token_id": "1",
      "valid_until": "20000",
      "memo": "test transaction"
    })),
    public_keys: None,
  };

  let response = mina_mesh.construction_metadata(request).await;

  assert!(response.is_err());
  assert_debug_snapshot!(response);
  Ok(())
}

#[tokio::test]
async fn construction_metadata_empty() -> Result<()> {
  let mina_mesh = MinaMeshConfig::from_env().to_mina_mesh().await?;
  let request = ConstructionMetadataRequest::new(network_id());
  let response = mina_mesh.construction_metadata(request).await;
  assert!(response.is_err());
  assert_debug_snapshot!(response);
  Ok(())
}

#[tokio::test]
async fn construction_metadata_missing_sender() -> Result<()> {
  let mina_mesh = MinaMeshConfig::from_env().to_mina_mesh().await?;
  let request = ConstructionMetadataRequest {
    network_identifier: network_id().into(),
    // cspell:disable-next-line
    options: Some(serde_json::json!({ "receiver": "B62qnXy1f75qq8c6HS2Am88Gk6UyvTHK3iSYh4Hb3nD6DS2eS6wZ4or" })),
    public_keys: None,
  };

  let response = mina_mesh.construction_metadata(request).await;

  assert!(response.is_err());
  assert_debug_snapshot!(response);
  Ok(())
}

#[tokio::test]
async fn construction_metadata_missing_receiver() -> Result<()> {
  let mina_mesh = MinaMeshConfig::from_env().to_mina_mesh().await?;
  let request = ConstructionMetadataRequest {
    network_identifier: network_id().into(),
    // cspell:disable-next-line
    options: Some(serde_json::json!({ "sender": "B62qnXy1f75qq8c6HS2Am88Gk6UyvTHK3iSYh4Hb3nD6DS2eS6wZ4or" })),
    public_keys: None,
  };

  let response = mina_mesh.construction_metadata(request).await;

  assert!(response.is_err());
  assert_debug_snapshot!(response);
  Ok(())
}

#[tokio::test]
async fn construction_metadata_missing_token_id() -> Result<()> {
  let mina_mesh = MinaMeshConfig::from_env().to_mina_mesh().await?;
  let request = ConstructionMetadataRequest {
    network_identifier: network_id().into(),
    options: Some(serde_json::json!({
      // cspell:disable
      "sender": "B62qnXy1f75qq8c6HS2Am88Gk6UyvTHK3iSYh4Hb3nD6DS2eS6wZ4or",
      "receiver": "B62qpLST3UC1rpVT6SHfB7wqW2iQgiopFAGfrcovPgLjgfpDUN2LLeg",
      // cspell:enable
    })),
    public_keys: None,
  };

  let response = mina_mesh.construction_metadata(request).await;

  assert!(response.is_err());
  assert_debug_snapshot!(response);
  Ok(())
}

#[tokio::test]
async fn construction_metadata_invalid_sender() -> Result<()> {
  let mina_mesh = MinaMeshConfig::from_env().to_mina_mesh().await?;
  let request = ConstructionMetadataRequest {
    network_identifier: network_id().into(),
    options: Some(serde_json::json!({
      "sender": "invalid",
      // cspell:disable-next-line
      "receiver": "B62qpLST3UC1rpVT6SHfB7wqW2iQgiopFAGfrcovPgLjgfpDUN2LLeg",
      "token_id": "1",
    })),
    public_keys: None,
  };

  let response = mina_mesh.construction_metadata(request).await;

  assert!(response.is_err());
  assert_debug_snapshot!(response);
  Ok(())
}

#[tokio::test]
async fn construction_metadata_invalid_receiver() -> Result<()> {
  let mina_mesh = MinaMeshConfig::from_env().to_mina_mesh().await?;
  let request = ConstructionMetadataRequest {
    network_identifier: network_id().into(),
    options: Some(serde_json::json!({
      // cspell:disable-next-line
      "sender": "B62qpLST3UC1rpVT6SHfB7wqW2iQgiopFAGfrcovPgLjgfpDUN2LLeg",
      "receiver": "invalid",
      "token_id": "1",
    })),
    public_keys: None,
  };

  let response = mina_mesh.construction_metadata(request).await;

  assert!(response.is_err());
  assert_debug_snapshot!(response);
  Ok(())
}
