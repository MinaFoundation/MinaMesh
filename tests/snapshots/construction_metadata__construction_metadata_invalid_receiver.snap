---
source: tests/construction_metadata.rs
expression: response
---
Err(
    JsonParse(
        Some(
            "Receiver key not valid: Malformed public key: Input not valid base58",
        ),
    ),
)
